zebra/zapi_msg.c

```diff
@@ -1319,7 +1319,31 @@ static void zread_interface_add(ZAPI_HANDLER_ARGS)
 	struct vrf *vrf;
 	struct interface *ifp;
 
-	RB_FOREACH (vrf, vrf_id_head, &vrfs_by_id) {
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s protocol: %s vrf_id: %d vrf_name: %s", __func__,
+			   zebra_route_string(client->proto), zvrf_id(zvrf),
+			   zvrf_name(zvrf));
+
+	 /* Interface information is needed. */
+	vrf_bitmap_set(client->ifinfo, zvrf_id(zvrf));
+
+	/* OSPF and BGP protocol daemons have per VRF registration for
+	 * interface events. Rest of the protocols register for default VRF
+	 * So, BGP and OSPF end up receicing interface events times the
+	 * no. of VRFs in the system. I tried to preserve the existing
+	 * behavior for other protocols
+	 */
+
+	if (client->proto == ZEBRA_ROUTE_BFD ||
```

Comment:
IMO this is just a bug waiting to happen with this setup.  No-one is going to remember to update this or the other place we do this magic.

Reply:
I agree with you. I intend to take care of this as another pull request. However I have done code changes to address this issue. With my latest change, All daemons will registers for all the VRFs. VRF aware daemons through configuration and others when the VRF is enabled. Similarly for the VRF unregister as well.